# TV Script Generation

In this project, you'll generate your own Seinfeld TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new, "fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/seinfeld_scripts.txt` and you're encouraged to open that file and look at the text. 
* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
data_dir = "./data/seinfeld_scripts.txt"
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [2]:
view_line_range = (3, 15)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print("Dataset Stats")
print("-------------\n")

print('Rough number of total words: {}'.format(len(text.split())))
print("Rough number of unique words: {}".format(len({word: None for word in text.split()})))

lines = text.split("\n")
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print("Average number of words per line: {:.3f}".format(np.average(word_count_line)))

print("Lines {} to {}:".format(*view_line_range))
print("\n".join(text.split("\n")[view_line_range[0]:view_line_range[1]]))

Dataset Stats
-------------

Rough number of total words: 605614
Rough number of unique words: 46367
Number of lines: 109233
Average number of words per line: 5.544
Lines 3 to 15:

george: are you through? 

jerry: you do of course try on, when you buy? 

george: yes, it was purple, i liked it, i dont actually recall considering the buttons. 

jerry: oh, you dont recall? 

george: (on an imaginary microphone) uh, no, not at this time. 

jerry: well, senator, id just like to know, what you knew and when you knew it. 


---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [3]:
from collections import Counter
import tester as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """

    # create lookup table
    words_count = Counter(text)
    vocab_to_int = {word: i for i, (word, count) in enumerate(words_count.most_common())}
    int_to_vocab = {i: word for word, i in vocab_to_int.items()}
    
    # return tuple
    return (vocab_to_int, int_to_vocab)

word2int, int2word = create_lookup_tables(text.split())

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [4]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # create token lookup table
    tokens = {
        ".": "<PERIOD>",
        ",": "<COMMA>",
        "\"": "<QUOTATION>",
        ";": "<SEMICOLON>",
        "!": "<EXCLAMATION>",
        "?": "<QUESTION>",
        "(": "<PARENTHESIS_LEFT>",
        ")": "<PARENTHESIS_RIGHT>",
        "-": "<DASH>",
        "\n": "<RETURN>",
    }
    return tokens

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to look at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import tester as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

In [7]:
print("Rough number of total words BEFORE pre-processing:", len(text.split()))
print("Rough number of total words AFTER pre-processing:", len(int_text))
print("Unique words: {}".format(len({word: None for word in int_text})))

Rough number of total words BEFORE pre-processing: 605614
Rough number of total words AFTER pre-processing: 892110
Unique words: 21387


## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

train_on_gpu = torch.cuda.is_available()

if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [9]:
from torch.utils.data import TensorDataset, DataLoader
import numpy as np
import torch

def batch_data(words, sequence_length, batch_size, shuffle=True):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    
    features = np.zeros((len(words)-sequence_length+1, sequence_length), dtype=int)
    targets = np.zeros(len(words)-sequence_length+1, dtype=int)
    
    for i in range(len(words)-sequence_length+1):
        features[i, :] = words[i:i+sequence_length]
        try:
            targets[i] = words[i+sequence_length]
        except IndexError:
            targets[i] = words[0]

    data = TensorDataset(torch.from_numpy(features), torch.from_numpy(targets))
    data_loader = DataLoader(data, shuffle=shuffle, batch_size=batch_size)
    
    return data_loader

data_loader = batch_data(range(50), 5, 10, shuffle=False)
data_iter = iter(data_loader)

for x, y in data_iter:
    print(x)
    print(y)

tensor([[ 0,  1,  2,  3,  4],
        [ 1,  2,  3,  4,  5],
        [ 2,  3,  4,  5,  6],
        [ 3,  4,  5,  6,  7],
        [ 4,  5,  6,  7,  8],
        [ 5,  6,  7,  8,  9],
        [ 6,  7,  8,  9, 10],
        [ 7,  8,  9, 10, 11],
        [ 8,  9, 10, 11, 12],
        [ 9, 10, 11, 12, 13]])
tensor([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])
tensor([[10, 11, 12, 13, 14],
        [11, 12, 13, 14, 15],
        [12, 13, 14, 15, 16],
        [13, 14, 15, 16, 17],
        [14, 15, 16, 17, 18],
        [15, 16, 17, 18, 19],
        [16, 17, 18, 19, 20],
        [17, 18, 19, 20, 21],
        [18, 19, 20, 21, 22],
        [19, 20, 21, 22, 23]])
tensor([15, 16, 17, 18, 19, 20, 21, 22, 23, 24])
tensor([[20, 21, 22, 23, 24],
        [21, 22, 23, 24, 25],
        [22, 23, 24, 25, 26],
        [23, 24, 25, 26, 27],
        [24, 25, 26, 27, 28],
        [25, 26, 27, 28, 29],
        [26, 27, 28, 29, 30],
        [27, 28, 29, 30, 31],
        [28, 29, 30, 31, 32],
        [29, 30, 31, 32, 33]])

In [106]:
data_loader = batch_data(int_text, 8, 10)
data_iter = iter(data_loader)
features, target = next(data_iter)
print(features)
print(target)

print()

for i, item in enumerate(features):
    print([int_to_vocab[j.item()] for j in item], int_to_vocab[int(target[i].numpy())])

tensor([[  31,   45,    1,    0,    0, 1706,  100,    2],
        [ 584,   15, 1332,  602,    1,    0,    0,   14],
        [   1,    0,    0,   14,  128,   84,   19,   93],
        [   1,   11,   21, 1004, 1855,   12, 5516,    2],
        [  17,    5,  204,   46,    8,   64,  134,  236],
        [7798,  104,   62,    1,    0,    0,   16,   11],
        [   1,    5, 1090,   37,  175,  112,  271,    8],
        [   0,   13,   37,   66,  536,   76,   38, 2598],
        [2833,   49,  177,  360,    2,  124,   77,    4],
        [  15,   49,   40,  477,  358,    1,    0,    0]])
tensor([    4,    56,    95, 14031,   298,  1601,    51,    10,    34,  1137])

["i'm", 'not', '<period>', '<return>', '<return>', 'patty:', 'yes', '<comma>'] you
['worry', 'it', "hasn't", 'started', '<period>', '<return>', '<return>', 'elaine:'] hey
['<period>', '<return>', '<return>', 'elaine:', 'where', 'did', 'that', 'come'] from
['<period>', '<parenthesis_left>', 'in', 'fake', 'spanish', '<parenthesis_right>', 

### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [11]:
test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 8,  9, 10, 11, 12],
        [ 0,  1,  2,  3,  4],
        [ 1,  2,  3,  4,  5],
        [14, 15, 16, 17, 18],
        [ 5,  6,  7,  8,  9],
        [45, 46, 47, 48, 49],
        [ 4,  5,  6,  7,  8],
        [12, 13, 14, 15, 16],
        [36, 37, 38, 39, 40],
        [24, 25, 26, 27, 28]])

torch.Size([10])
tensor([13,  5,  6, 19, 10,  0,  9, 17, 41, 29])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [12]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5, batch_first=True):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        :param batch_first: first dimension of input and output tensors is the batch size
        """
        super(RNN, self).__init__()
        
        # set class variables
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        self.output_size = output_size
        
        # define model layers
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout=dropout, batch_first=batch_first)
        self.drop = nn.Dropout(p=0.5)
        self.fc = nn.Linear(hidden_dim, output_size)
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """

        # embeddings and lstm
        embed_out = self.embedding(nn_input)
        lstm_out, hidden = self.lstm(embed_out, hidden)
        
        # dropout
        out = self.drop(lstm_out)
        
        # stack up the LSTM outputs to pass to last linear layer
        out = out.contiguous().view(-1, self.hidden_dim)
        
        # fully-connected layer
        out = self.fc(out)
        
        # get the last batch of word scores
        out = out.view(nn_input.size(0), -1, self.output_size)
        output = out[:, -1]
        
        # return one batch of output word scores and the hidden state
        return output, hidden
        
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        
        weight = next(self.parameters()).data
        if train_on_gpu:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


In [13]:
# instantiate a test model
rnn_test = RNN(100, 100, 200, 300, 2, dropout=0.5)
train_loader = batch_data(range(50), 10, 3)
data_iter = iter(train_loader)
x, y = data_iter.next()
print(x)
out, _ = rnn_test.forward(x, None)
print(out.shape)

tensor([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10],
        [27, 28, 29, 30, 31, 32, 33, 34, 35, 36],
        [ 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]])
torch.Size([3, 100])


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and backpropagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [14]:
def forward_back_prop(rnn, optimizer, criterion, features, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param features: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # move data to GPU
    if(train_on_gpu):
        features, target = features.cuda(), target.cuda()
    
    # detach from history
    hidden = tuple([each.data for each in hidden])

    # zero accumulated gradients
    rnn.zero_grad()
    
    # get the output from the model
    output, h = rnn(features, hidden)
    
    # perform backpropagation and optimization
    loss = criterion(output, target)
    loss.backward()

    # "clip_grad_norm" helps prevent the exploding gradient problem in RNNs / LSTMs.
    nn.utils.clip_grad_norm_(rnn.parameters(), 0.5)
    optimizer.step()

    # return the loss over a batch and the hidden state
    return loss.item(), hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()
    
    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

        # printing loss stats
        print("Epoch: {}/{}   Loss: {}".format(epoch_i, n_epochs, np.average(batch_losses)))
        batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of unique tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [16]:
## Data Parameters
# Sequence Length
sequence_length = 5
# Batch Size
batch_size = 512

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [17]:
## Training Parameters
# Number of Epochs
num_epochs = 20
# Learning Rate
learning_rate = 0.001

## Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = len(vocab_to_int)
# Embedding Dimension
embedding_dim = 200
# Hidden Dimension
hidden_dim = 400
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 1000

### Train
In the next cell, you'll train the neural network on the pre-processed data. If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and layer dimensions, but larger models take a longer time to train. 

**You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/model_trained', trained_rnn)
print('Model Trained and Saved')

Training for 20 epoch(s)...
Epoch: 1/20   Loss: 4.682519107543779
Epoch: 2/20   Loss: 4.161264402036963
Epoch: 3/20   Loss: 4.003129777213052
Epoch: 4/20   Loss: 3.9009675523854286
Epoch: 5/20   Loss: 3.8211219915703163
Epoch: 6/20   Loss: 3.7547311353081265
Epoch: 7/20   Loss: 3.699236129099411
Epoch: 8/20   Loss: 3.6518289693871813
Epoch: 9/20   Loss: 3.6089666216846994
Epoch: 10/20   Loss: 3.5703928020326194
Epoch: 11/20   Loss: 3.534844255337896
Epoch: 12/20   Loss: 3.50342281965109
Epoch: 13/20   Loss: 3.474121647642072
Epoch: 14/20   Loss: 3.4464798829038163
Epoch: 15/20   Loss: 3.420058573428306
Epoch: 16/20   Loss: 3.396588951930934
Epoch: 17/20   Loss: 3.3736607278392454
Epoch: 18/20   Loss: 3.3514895614334965
Epoch: 19/20   Loss: 3.3302480614691734
Epoch: 20/20   Loss: 3.312414601120144
Model Trained and Saved


### Question: How did you decide on your model hyperparameters? 
For example, did you try different `sequence_lengths` and find that one size made the model converge faster? What about your `hidden_dim` and `n_layers`, how did you decide on those?

**Answer:**
I have found out that a small learning rate of `lr = 0.001` and a large batch size of `batch_size = 512` made the model converge the fastest. I had to try a number of different values for the batch size to get the model to train properly. I have started with a value of 50, but training turned out to be too slow in this range. I decided to choose a sequence length of `sequence_length = 5`, because the average number of words per line is `5.544` according to the dataset statistics in the beginning of this project. I felt there was no need to go with a higher value, since I wanted to generate a script that has similar complexity as the original content. Since the dictionary contains a number of 21387 unique words, I chose a value of `embedding_dim = 200` for the embedding dimension. This is based on my experience of previous RNN implementations. For the number of hidden features, `hidden_dim = 400` turned out to be a good value to achieve enough learning capacity. I wanted to balance model performance vs. training time. In order to keep training time low, I set the number of LSTM layers to `n_layers = 2`.

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import tester as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/model_trained')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        if(train_on_gpu):
            current_seq = current_seq.cpu() # move to cpu
        # the generated word becomes the next "current sequence" and the cycle can continue
        if train_on_gpu:
            current_seq = current_seq.cpu()
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [113]:
# run the cell multiple times to get different results!
gen_length = 200
prime_word = 'jerry'

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

jerry:) hey!

jerry: i can't believe this, i was in the hospital to be a little more tactful.

jerry:(to jerry) hey, you know...(he leaves)

george: you know what? you know, i have a very bad idea.

george: what is it?!!

kramer:(to jerry) you got the job?

helen: i don't know why we had a little more time.

jerry: you know what, i think i can see the whole thing, and i was thinking of a little problem with you?

george:(looking at the tv)

jerry: i think you're going to go.(to george) hey, you gotta see the doctor?

elaine: no, no. no.

newman:(to jerry) you know i don't know what you think.

kramer: hey, you know....

kramer: oh yeah, you know i was gonna get out


#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [114]:
f = open("script.txt","w")
f.write(generated_script)
f.close()

### Discuss the New Script

You can see that there are multiple characters that say somewhat complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary and discard uncommon words, or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally.